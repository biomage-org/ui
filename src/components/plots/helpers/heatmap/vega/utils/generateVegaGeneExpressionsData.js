const cartesian = (...array) => (
  array.reduce((acum, value) => (
    acum.flatMap((d) => (
      value.map((e) => [d, e].flat())
    ))
  ))
);

const generateVegaGeneExpressionsData = (cellOrder, geneOrder, expression, heatmapSettings) => {
  const { expressionValue, truncatedValues } = heatmapSettings;

  const geneExpressionsData = [];

  if (!expression.matrix.genesAreLoaded(geneOrder)) {
    return;
  }

  // Preload all genes so that their arrays are generated only once
  const preloadedExpressions = {};
  geneOrder.forEach((gene) => {
    const geneExpression = { rawExpression: expression.matrix.getRawExpression(gene, cellOrder) };
    geneExpression.truncatedExpression = expression.matrix.getTruncatedExpression(gene, cellOrder);

    preloadedExpressions[gene] = geneExpression;
  });

  const cellOrderWithIndexes = cellOrder.map((cellId, index) => ({ cellId, index }));

  cartesian(geneOrder, cellOrderWithIndexes).forEach(
    ([gene, { cellId, index }]) => {
      const expressionValues = {};

      if (expressionValue === 'zScore') {
        throw new Error('Not implemented DO NOT MERGE: Zscore needs to be implemented for sparse matrix in the worker first');
        // expressionValues = {
        //   color: expressionDataForGene.zScore, display: expressionDataForGene.zScore,
        // };
      } else {
        expressionValues.display = preloadedExpressions[gene].rawExpression;
        expressionValues.color = truncatedValues
          ? preloadedExpressions[gene].truncatedExpression
          : preloadedExpressions[gene].display;
      }

      geneExpressionsData.push({
        cellId,
        gene,
        expression: expressionValues.color[index],
        displayExpression: expressionValues.display[index],
      });
    },
  );

  return geneExpressionsData;
};

export default generateVegaGeneExpressionsData;
