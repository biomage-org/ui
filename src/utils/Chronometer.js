import moment from 'moment';

class Chronometer {
  static chronometers = {};

  /**
   *
   * @param {*} key Key of the chronometer to get
   * @returns
   */
  static get(key) {
    return this.chronometers[key];
  }

  /**
   *
   * @param {*} key Optional. If set, it will make this chronometer
   * accessible through Chronometer.get(key)
   */
  constructor(key = undefined) {
    if (key) {
      this.key = key;

      Chronometer.chronometers[key] = this;
    }
  }

  start() {
    this.startTime = moment();
  }

  stop() {
    return moment.duration(moment().diff(this.startDate));
  }
}

export default Chronometer;
