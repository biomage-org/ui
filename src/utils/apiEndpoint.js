const getApiEndpoint = (location) => {
  try {
    const url = new URL(location || window.location.href);

    if (url.hostname.includes('staging')) {
      return url.origin.replace('ui', 'api');
    }

    if (url.hostname.includes('localhost') || url.hostname.includes('127.0.0.1')) {
      return 'http://localhost:3000';
    }

    return `https://api.${process.env.DOMAIN_NAME}`;
  } catch (error) {
    console.error('Failed to get API endpoint.');
    console.error(error);
  }
};

export default getApiEndpoint;
