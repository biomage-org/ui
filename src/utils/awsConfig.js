import * as AWS from '@aws-sdk/client-sts';
import { getDefaultRoleAssumerWithWebIdentity } from '@aws-sdk/client-sts';
import { fromTokenFile } from '@aws-sdk/credential-provider-web-identity';

const getAccountId = async () => {
  const sts = new AWS.STS({
    region: getAWSRegion(),
    credentials: fromTokenFile({
      roleAssumerWithWebIdentity: getDefaultRoleAssumerWithWebIdentity(),
    }),
  });
  console.log('HERE LMAO', getAWSRegion());

  const accountID = await sts.getCallerIdentity({}).promise;
  console.log('ID IS ', accountID);

  return accountID;
};

const getAWSRegion = () => {
  const region = process.env.AWS_REGION || process.env.DEFAULT_REGION || 'eu-west-1';
  return region;
};

export { getAccountId, getAWSRegion };
