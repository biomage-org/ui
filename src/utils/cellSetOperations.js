const unionByCellClass = (cellClassKey, hierarchy, properties) => {
  const cellSetKeys = hierarchy
    .find(({ key }) => key === cellClassKey).children
    .map(({ key }) => key);

  return union(cellSetKeys, properties);
};

const union = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }
  const sets = listOfSets.map((key) => properties[key]?.cellIds || []);
  const unionSet = new Set(
    [].concat(
      ...sets.map(
        (set) => [...set],
      ),
    ),
  );

  return unionSet;
};

const intersection = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }

  const sets = listOfSets.map(
    (key) => properties[key]?.cellIds || null,
  ).filter(
    (set) => set && set.size > 0,
  );

  if (sets.length === 0) {
    return new Set();
  }

  const intersectionSet = sets.reduce(
    (acc, curr) => new Set([...acc].filter((x) => curr.has(x))),
  );

  return intersectionSet;
};

const complement = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }

  // get the ids of all selected cells
  const selectedCells = listOfSets.map(
    (key) => properties[key]?.cellIds || null,
  ).filter(
    (set) => set && set.size > 0,
  ).reduce(
    (acc, curr) => new Set([...acc, ...curr]),
  );

  // get the ids of all cells in the dataset
  // All cells are assumed to be included in at least 1 cluster
  const complementSet = Object.values(properties).map(
    (cluster) => cluster.cellIds,
  ).filter(
    (set) => set && set.size > 0,
  ).reduce(
    (acc, curr) => new Set([...acc, ...curr]),
  );

  // remove all cells that are selected
  if (selectedCells.size > 0) {
    selectedCells.forEach((x) => { complementSet.delete(x); });
  }

  // return the rest of the cells that are in the dataset and were not selected
  return complementSet;
};

export {
  union,
  intersection,
  complement,
  unionByCellClass,
};
