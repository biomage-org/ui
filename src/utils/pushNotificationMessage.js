import { message } from 'antd';

const pushNotificationMessage = (type, text) => {
  switch (type) {
    case 'success':
      message.success(text, 2);
      break;
    case 'error':
      message.error(text, 4);
      break;
    case 'info':
      message.info(text, 4);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, 4);
      break;
    case 'loading':
      message.loading(text, 2);
      break;
    default:
      message.info(text, 4);
      break;
  }
};

export default pushNotificationMessage;
