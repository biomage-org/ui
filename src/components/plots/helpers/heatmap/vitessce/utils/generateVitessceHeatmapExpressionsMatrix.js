import _ from 'lodash';
import { convertRange } from 'utils/plotUtils';

const scaledTo255 = (rowOfExpressions) => {
  const min = _.min(rowOfExpressions);
  const max = _.max(rowOfExpressions);

  return rowOfExpressions.map((value) => convertRange(value, [min, max], [0, 255]));
};

const generateVitessceHeatmapExpressionsMatrix = (cellOrder, geneOrder, expression) => {
  const geneExpressionsDataMatrix = [];

  geneOrder.forEach((gene) => {
    const { matrix } = expression;

    if (!matrix.geneIsLoaded(gene)) return;

    console.log('[DEBUG] - BEGUN matrix.getTruncatedExpression(gene)');
    const truncatedExpression = matrix.getTruncatedExpression(gene);
    console.log('[DEBUG] - FINISHED matrix.getTruncatedExpression(gene)');

    console.log('[DEBUG] - BEGUN cellOrder.map((cellId) => truncatedExpression[cellId])');
    const geneExpressions = cellOrder.map((cellId) => truncatedExpression[cellId]);
    console.log('[DEBUG] - FINISHED cellOrder.map((cellId) => truncatedExpression[cellId])');

    console.log('[DEBUG] - BEGUN scaledTo255');
    const scaledGeneExpressions = scaledTo255(geneExpressions);
    console.log('[DEBUG] - FINISHED scaledTo255');

    geneExpressionsDataMatrix.push(scaledGeneExpressions);
  });

  const cellExpressionsData = _.flatten(_.unzip(geneExpressionsDataMatrix));

  return cellExpressionsData;
};

export default generateVitessceHeatmapExpressionsMatrix;
