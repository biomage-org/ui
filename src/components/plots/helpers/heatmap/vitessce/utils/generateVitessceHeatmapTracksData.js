import getContainingCellSetsProperties from 'utils/cellSets/getContainingCellSetsProperties';
import { hexToRgb } from 'utils/plotUtils';

const generateVitessceHeatmapTracksData = (trackOrder, cellSets, cells) => {
  const colorForCell = (cellId, cellClassKey) => {
    const { color: cellColor = null } = getContainingCellSetsProperties(
      cellId,
      [cellClassKey],
      cellSets,
    )[cellClassKey][0] ?? {};

    return hexToRgb(cellColor) ?? hexToRgb('#f5f8fa');
  };

  const cellIdsColorsMap = new Map();

  cells.forEach((cellId) => {
    const allColorsForCell = trackOrder.map((cellClassKey) => colorForCell(cellId, cellClassKey));

    cellIdsColorsMap.set(`${cellId}`, allColorsForCell);
  });

  return cellIdsColorsMap;
};

export default generateVitessceHeatmapTracksData;
