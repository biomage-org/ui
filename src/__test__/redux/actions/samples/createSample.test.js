import _ from 'lodash';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 as uuidv4 } from 'uuid';

import createSample from 'redux/actions/samples/createSample';
import initialSampleState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';
import 'utils/upload/validate10x';

import {
  SAMPLES_CREATE, SAMPLES_SAVING, SAMPLES_ERROR, SAMPLES_SAVED,
} from 'redux/actionTypes/samples';

import endUserMessages from 'utils/endUserMessages';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import { sampleTech } from 'utils/constants';

jest.mock('utils/upload/validate10x');
jest.mock('utils/upload/validateRhapsody');
pushNotificationMessage.mockImplementation(() => async () => { });

enableFetchMocks();

const mockStore = configureStore([thunk]);

jest.mock('uuid');
const sampleUuid = 'abc123';
uuidv4.mockImplementation(() => sampleUuid);

const sampleName = 'test sample';
const sample = {};

describe('createSample action', () => {
  const experimentId = 'exp234';

  const mockType = sampleTech['10X'];

  const mockExperiment = {
    ...experimentTemplate,
    name: 'Experiment 1',
    id: experimentId,
    metadataKeys: ['meta-1', 'meta-2'],
  };

  const initialState = {
    samples: { ...initialSampleState },
    experiments: {
      ...initialExperimentState,
      [experimentId]: mockExperiment,
      ids: [mockExperiment.id],
    },
  };

  let store;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();

    store = mockStore(initialState);
  });

  it('Works correctly with one file being uploaded', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(experimentId, sampleName, sample, mockType, ['matrix.tsv.gz']));

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v2/experiments/${mockExperiment.id}/samples/${sampleUuid}`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_CREATE, SAMPLES_SAVED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Works correctly with many files being uploaded', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(experimentId, sampleName, sample, mockType, ['matrix.tsv.gz', 'features.tsv.gz', 'barcodes.tsv.gz']));

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v2/experiments/${mockExperiment.id}/samples/${sampleUuid}`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_CREATE, SAMPLES_SAVED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Throws if the api fails', async () => {
    fetchMock.mockRejectOnce(() => Promise.reject(new Error('Some error')));

    await expect(
      store.dispatch(
        createSample(experimentId, sampleName, sample, mockType, ['matrix.tsv.gz']),
      ),
    ).rejects.toThrow(endUserMessages.ERROR_CREATING_SAMPLE);

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_ERROR]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Throws if technology is not identified', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    await expect(
      store.dispatch(
        createSample(experimentId, sampleName, sample, 'unrecognizable type', ['matrix.tsv.gz', 'features.tsv.gz', 'barcodes.tsv.gz']),
      ),
    ).rejects.toThrow('Sample technology unrecognizable type is not recognized');
  });

  it('Works correctly with BDRhapsody file being uploaded', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(experimentId, sampleName, sample, sampleTech.RHAPSODY, ['asdasda_Expression_Data.st']));

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v2/experiments/${mockExperiment.id}/samples/${sampleUuid}`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_CREATE, SAMPLES_SAVED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Creates sample with options value that is the same as existing samples', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const existingSampleId = 'mock-sample';
    const mockOptions = { someOption: true, otherOption: false };

    const existingSampleState = {
      ...initialState,
      experiments: {
        ...initialState.experiments,
        [experimentId]: {
          ...initialState.experiments[experimentId],
          sampleIds: [existingSampleId],
        },
      },
      samples: {
        ...initialSampleState,
        [existingSampleId]: {
          ...sampleTemplate,
          options: mockOptions,
        },
      },
    };

    store = mockStore(existingSampleState);

    const newUuid = await store.dispatch(
      createSample(experimentId, sampleName, sample, sampleTech.RHAPSODY, ['asdasda_Expression_Data.st']),
    );

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody } = fetchMockFirstCall[1];

    const payload = JSON.parse(fetchBody);

    expect(payload.options).toEqual(mockOptions);
    expect(payload).toMatchSnapshot();
  });
});
