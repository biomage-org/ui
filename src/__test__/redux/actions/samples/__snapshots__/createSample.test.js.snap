// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSample action Creates sample with options value that is the same as existing samples 1`] = `
Object {
  "name": "test sample",
  "options": Object {
    "otherOption": false,
    "someOption": true,
  },
  "sampleTechnology": "rhapsody",
}
`;

exports[`createSample action Throws if the api fails 1`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "error": "We couldn't create your sample. Please try uploading it again",
  },
]
`;

exports[`createSample action Works correctly with BDRhapsody file being uploaded 1`] = `
Object {
  "name": "test sample",
  "options": Object {
    "includeAbSeq": false,
  },
  "sampleTechnology": "rhapsody",
}
`;

exports[`createSample action Works correctly with BDRhapsody file being uploaded 2`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "experimentId": "exp234",
      "fileNames": Array [],
      "files": Object {
        "asdasda_Expression_Data.st": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {
        "meta-1": "N.A.",
        "meta-2": "N.A.",
      },
      "name": "test sample",
      "options": Object {
        "includeAbSeq": false,
      },
      "type": "rhapsody",
      "uuid": "abc123",
    },
  },
  undefined,
]
`;

exports[`createSample action Works correctly with many files being uploaded 1`] = `
Object {
  "name": "test sample",
  "options": Object {},
  "sampleTechnology": "10x",
}
`;

exports[`createSample action Works correctly with many files being uploaded 2`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "experimentId": "exp234",
      "fileNames": Array [],
      "files": Object {
        "barcodes.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "features.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {
        "meta-1": "N.A.",
        "meta-2": "N.A.",
      },
      "name": "test sample",
      "options": Object {},
      "type": "10x",
      "uuid": "abc123",
    },
  },
  undefined,
]
`;

exports[`createSample action Works correctly with one file being uploaded 1`] = `
Object {
  "name": "test sample",
  "options": Object {},
  "sampleTechnology": "10x",
}
`;

exports[`createSample action Works correctly with one file being uploaded 2`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "experimentId": "exp234",
      "fileNames": Array [],
      "files": Object {
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {
        "meta-1": "N.A.",
        "meta-2": "N.A.",
      },
      "name": "test sample",
      "options": Object {},
      "type": "10x",
      "uuid": "abc123",
    },
  },
  undefined,
]
`;
